src/elliptic/curves/p256.rs

```diff
@@ -0,0 +1,889 @@
+pub type FE = Secp256r1Scalar;
+
+/* X coordinate of a point of unknown discrete logarithm.
+Computed using a deterministic algorithm with the generator as input.
+See test_base_point2 */
+const BASE_POINT2_X: [u8; 32] = [
+    0x70, 0xf7, 0x2b, 0xba, 0xc4, 0x0e, 0x8a, 0x59, 0x4c, 0x91, 0xa7, 0xba, 0xc3, 0x76, 0x59, 0x27, 0x89, 0x10, 0x76, 0x4c, 0xd7, 0xc2, 0x0a, 0x7d, 0x65, 0xa5, 0x9a, 0x04, 0xb0, 0xac, 0x2a, 0xde
+];
+const BASE_POINT2_Y: [u8; 32] = [
+    0x30, 0xe2, 0xfe, 0xb3, 0x8d, 0x82, 0x4e, 0x0e, 0xa2, 0x95, 0x2f, 0x2a, 0x48, 0x5b, 0xbc, 0xdd, 0x4c, 0x72, 0x8a, 0x74, 0xf4, 0xfa, 0xc7, 0xdc, 0x0d, 0xc9, 0x90, 0x8d, 0x9a, 0x8d, 0xc1, 0xa4
+];
+
+impl Zeroize for Secp256r1Scalar {
+    fn zeroize(&mut self) {
+        unsafe { ptr::write_volatile(self, Secp256r1Scalar::zero()) };
+        atomic::fence(atomic::Ordering::SeqCst);
+        atomic::compiler_fence(atomic::Ordering::SeqCst);
+    }
+}
+
+impl ECScalar<SK> for Secp256r1Scalar {
+    fn new_random() -> Secp256r1Scalar {
+        let mut arr = [0u8; 32];
+        thread_rng().fill(&mut arr[..]);
+        Secp256r1Scalar {
+            purpose: "random",
+            fe: SK::from_bytes(&arr).unwrap(),
+        }
+    }
+
+    fn zero() -> Secp256r1Scalar {
+        let zero_arr = [0u8; 32];
+        let zero = unsafe { std::mem::transmute::<[u8; 32], SecretKey>(zero_arr) };
+        Secp256r1Scalar {
+            purpose: "zero",
+            fe: zero,
+        }
+    }
+
+    fn get_element(&self) -> SK {
+        self.fe.clone()
+    }
+
+    fn set_element(&mut self, element: SK) {
+        self.fe = element
+    }
+
+    fn from(n: &BigInt) -> Secp256r1Scalar {
+        let curve_order = Secp256r1Scalar::q();
+        let n_reduced = BigInt::mod_add(n, &BigInt::from(0), &curve_order);
+        let mut v = BigInt::to_vec(&n_reduced);
+        const SECRET_KEY_SIZE: usize = 32;
+
+        if v.len() < SECRET_KEY_SIZE {
+            let mut template = vec![0; SECRET_KEY_SIZE - v.len()];
+            template.extend_from_slice(&v);
+            v = template;
+        }
+
+        Secp256r1Scalar {
+            purpose: "from_big_int",
+            fe: SK::from_bytes(&v).unwrap(),
+        }
+    }
+
+    fn to_big_int(&self) -> BigInt {
+        BigInt::from(self.fe.to_bytes().as_slice())
+    }
+
+    fn q() -> BigInt {
+        const CURVE_ORDER: [u8; 32] = [
+            0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+            0xff, 0xff, 0xbc, 0xe6, 0xfa, 0xad, 0xa7, 0x17, 0x9e, 0x84, 0xf3, 0xb9, 0xca, 0xc2,
+            0xfc, 0x63, 0x25, 0x51,
+        ];
+        BigInt::from(CURVE_ORDER.as_ref())
+    }
+
+    fn add(&self, other: &SK) -> Secp256r1Scalar {
+        let scalar1 = Scalar::from_bytes_reduced(&self.get_element().to_bytes());
+        let scalar2 = Scalar::from_bytes_reduced(&other.to_bytes());
+        Secp256r1Scalar {
+            purpose: "add",
+            fe: SK::new(NonZeroScalar::new(scalar1 + scalar2).unwrap()),
```

Comment:
Can't `scalar1 + scalar2` be zero?
If so, is it still ok to use `NonZeroScalar`?

Reply:
Good catch! That particular issue is "solved" in https://github.com/ZenGo-X/curv/pull/87/commits/e910270f0facac3e6e23c793dd9038512d130822, but there is just no check anymore whether the result is 0..